client/sysinfo/src/sysinfo.rs

```diff
@@ -32,14 +33,113 @@ use std::{
 	time::{Duration, Instant},
 };
 
+pub enum Unit {
+	GiBs,
+	MiBs,
+	KiBs,
+}
+
+impl fmt::Display for Unit {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		let unit = match self {
+			Unit::GiBs => "GiBs",
+			Unit::MiBs => "MiBs",
+			Unit::KiBs => "KiBs",
+		};
+		write!(f, "{}", unit)
+	}
+}
+
+/// Throughput as measured in bytes per second.
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
+pub struct Throughput(f64);
+
+const KIBIBYTE: f64 = 1024.0;
+
+impl Throughput {
+	/// `f64` kibibyte/s to byte/s.
+	pub fn from_kibs(kibs: f64) -> Throughput {
+		Throughput(kibs * KIBIBYTE)
+	}
+
+	/// `f64` mebibyte/s to byte/s.
+	pub fn from_mibs(mibs: f64) -> Throughput {
+		Throughput(mibs * KIBIBYTE * KIBIBYTE)
+	}
+
+	/// `f64` gibibyte/s to byte/s.
+	pub fn from_gibs(gibs: f64) -> Throughput {
+		Throughput(gibs * KIBIBYTE * KIBIBYTE * KIBIBYTE)
+	}
+
+	pub fn as_f64(&self) -> f64 {
+		self.0
+	}
+
+	/// [`Self`] as number of kibibyte/s.
+	pub fn as_kibs(&self) -> f64 {
+		self.0 / KIBIBYTE
+	}
+
+	/// [`Self`] as number of mebibyte/s.
+	pub fn as_mibs(&self) -> f64 {
+		self.0 / (KIBIBYTE * KIBIBYTE)
+	}
+
+	/// [`Self`] as number of gibibyte/s.
+	pub fn as_gibs(&self) -> f64 {
+		self.0 / (KIBIBYTE * KIBIBYTE * KIBIBYTE)
+	}
+
+	/// Normalizes [`Self`] to use the larges unit possible.
+	pub fn normalize(&self) -> (f64, Unit) {
+		let bs = self.0;
+
+		if bs >= KIBIBYTE * KIBIBYTE * KIBIBYTE {
+			(self.as_gibs(), Unit::GiBs)
+		} else if bs >= KIBIBYTE * KIBIBYTE {
+			(self.as_mibs(), Unit::MiBs)
+		} else {
+			(self.as_kibs(), Unit::KiBs)
+		}
+	}
+}
+
+impl fmt::Display for Throughput {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		let (value, unit) = self.normalize();
+		write!(f, "{:.2?} {}", value, unit.to_string())
+	}
+}
+
+pub fn serialize_throughput_as_mibs<S>(t: &Throughput, serializer: S) -> Result<S::Ok, S::Error>
+where
+	S: Serializer,
+{
+	serializer.serialize_u64(t.as_mibs() as u64)
+}
+
+pub fn serialize_throughput_option_as_mibs<S>(
+	maybe_t: &Option<Throughput>,
+	serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+	S: Serializer,
+{
+	if let Some(t) = maybe_t {
+		return serializer.serialize_u64(t.as_mibs() as u64)
```

Comment:
Nitpick: in case of json it doesn't change anything, but this should be using `serialize_some`, as shown in the example snippet to which I linked to in my previous comment.